### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by rahularcota1**

### CODE REVIEW

File: app.go [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+type App struct {
+	Router *mux.Router
+	DB     *sql.DB
+}
+
+// tom: initial function is empty, it's filled afterwards
+// func (a *App) Initialize(user, password, dbname string) { }
+
+// tom: added "sslmode=disable" to connection string
+func (a *App) Initialize(user, password, dbname string) {
```
Comment (ayushaga12) [inline app.go]: don't keep everything in app.go. As we discussed, please break these down into separate packages. For ex - have a separate package for the api handler, models. Have a separate db connection init file, and use that db object for querying

File: app.go [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+	connectionString :=
+		fmt.Sprintf("user=%s password=%s dbname=%s sslmode=disable", user, password, dbname)
+		//fmt.Sprintf("user=%s password=%s dbname=%s", user, password, dbname)
+
+	var err error
+	a.DB, err = sql.Open("postgres", connectionString)
```
Comment (ayushaga12) [inline app.go]: use gorm

File: model.go [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+// tom: add backticks to json
+type product struct {
+	ID    int     `json:"id"`
+	Name  string  `json:"name"`
+	Price float64 `json:"price"`
+}
+
+// tom: these are added after tdd tests
+func (p *product) getProduct(db *sql.DB) error {
+	return db.QueryRow("SELECT name, price FROM products WHERE id=$1",
```
Comment (ayushaga12) [inline model.go]: use gorm everywhere

File: model.go [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+		p.Name, p.Price).Scan(&p.ID)
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func getProducts(db *sql.DB, start, count int) ([]product, error) {
```
Comment (ayushaga12) [inline model.go]: not required, maybe for now we can have a single api for fetching all products in the store

File: models/store.go [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+	Name string `json:"name"`
+}
+
+func(s *Store) CreateStore(db *gorm.DB) error {
+	txn := db.Create(&s)
+
+	return txn.Error
+}
+
+func(s *Store) CheckStoreExistence(db *gorm.DB) (bool, error){
```
Comment (ayushaga12) [inline models/store.go]: rename this to getStore

File: main_test.go [hunk 6]
```diff
@@ -0,0 +1,357 @@
+package main
+
+import (
```
Comment (ayushaga12) [inline main_test.go]: move all these tests inside a suite as discussed

File: models/product.go [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+	//return db.QueryRow("SELECT name, price FROM products WHERE id=$1",
+	//	p.ID).Scan(&p.Name, &p.Price)
+}
+
+func (p *Product) UpdateProduct(db *gorm.DB) error {
+	//_, err :=
+	//	db.Exec("UPDATE products SET name=$1, price=$2 WHERE id=$3",
+	//		p.Name, p.Price, p.ID)
+	//
+	//return err
```
Comment (shivamgor123) [inline models/product.go:27]: these functions should be in separate dao, not in models.

File: models/product.go [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+	txn := db.Delete(&p, p.ID)
+	return txn.Error
+}
+
+func (p *Product) CreateProduct(db *gorm.DB) error {
+	//err := db.QueryRow(
+	//	"INSERT INTO products(name, price) VALUES($1, $2) RETURNING id",
+	//	p.Name, p.Price).Scan(&p.ID)
```
Comment (shivamgor123) [inline models/product.go]: these functions should be in separate dao, not in models.

File: dal/models/product.go [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+	gorm.Model
+
+	Name  string  `json:"name"`
+	Price float64 `json:"price"`
+}
+
+//func (p *Product) GetProduct(db *gorm.DB) error {
+//	txn := db.First(&p, p.ID)
+//	return txn.Error
+//}
```
Comment (shivamgor123) [inline dal/models/product.go]: remove commented code, rest looks good to me

**Review by shivamgor123 — APPROVED** [review]

**Pull request merged**